provider/reprovider.go

```diff
@@ -228,216 +247,143 @@ func Online(rsys Provide) Option {
-				if len(m) == 0 {
-					noReprovideInFlight = s.noReprovideInFlight
+	} else {
+		provideQueue := make(chan cid.Cid)
+		provideDelayTimer := time.NewTimer(provideDelayWarnDuration)
+		provideDelayTimer.Stop()
+		lateOnProvides := false
+
+		// Assign cid to workers pool
+		provideOperation = func(ctx context.Context, c cid.Cid) {
+			provideDelayTimer.Reset(provideDelayWarnDuration)
+			defer provideDelayTimer.Stop()
+			select {
+			case provideQueue <- c:
+				if lateOnProvides {
+					log.Warn("New provides are being processed again.")
 				}
-
-				select {
-				case c := <-provCh:
-					resetTimersAfterReceivingProvide()
-					m[c] = struct{}{}
-				case c := <-s.reprovideCh:
-					resetTimersAfterReceivingProvide()
-					m[c] = struct{}{}
-					performedReprovide = true
-				case <-pauseDetectTimer.C:
-					// If this timer has fired then the max collection timer has started, so stop it.
-					maxCollectionDurationTimer.Stop()
-					complete = true
-					goto ProcessBatch
-				case <-maxCollectionDurationTimer.C:
-					// If this timer has fired then the pause timer has started, so stop it.
-					pauseDetectTimer.Stop()
-					goto ProcessBatch
-				case <-s.ctx.Done():
-					return
-				case noReprovideInFlight <- struct{}{}:
-					// If no reprovide is in flight get consumer asking for reprovides unstuck.
+				lateOnProvides = false
+			case <-provideDelayTimer.C:
+				if !lateOnProvides {
+					log.Warn("New provides are piling up in the queue, consider increasing the number of provide workers.")
+					lateOnProvides = true
 				}
-			}
-			pauseDetectTimer.Stop()
-			maxCollectionDurationTimer.Stop()
-		ProcessBatch:
-
-			if len(m) == 0 {
-				continue
-			}
-
-			keys := make([]multihash.Multihash, 0, len(m))
-			for c := range m {
-				delete(m, c)
-
-				// hash security
-				if err := verifcid.ValidateCid(s.allowlist, c); err != nil {
-					log.Errorf("insecure hash in reprovider, %s (%s)", c, err)
-					continue
+				select {
+				case provideQueue <- c:
+				case <-ctx.Done():
 				}
-
-				keys = append(keys, c.Hash())
-			}
-
-			// in case after removing all the invalid CIDs there are no valid ones left
-			if len(keys) == 0 {
-				continue
+			case <-ctx.Done():
 			}
-
-			if r, ok := s.rsys.(Ready); ok {
-				ticker := time.NewTicker(time.Minute)
-				for !r.Ready() {
-					log.Debugf("reprovider system not ready")
+		}
+		// Start provide workers
+		for range s.provideWorkerCount {
+			go func(ctx context.Context) {
+				for {
 					select {
-					case <-ticker.C:
-					case <-s.ctx.Done():
+					case c := <-provideQueue:
+						provideFunc(ctx, c)
+					case <-ctx.Done():
 						return
 					}
 				}
-				ticker.Stop()
-			}
+			}(s.ctx)
+		}
+	}
 
-			log.Debugf("starting provide of %d keys", len(keys))
-			start := time.Now()
-			err := doProvideMany(s.ctx, s.rsys, keys)
-			if err != nil {
-				log.Debugf("providing failed %v", err)
+	for {
+		select {
+		case c := <-provCh:
```

Comment:
Need to check if channel closed, because of [this](https://github.com/ipfs/boxo/blob/main/provider/internal/queue/queue.go#L87).
```suggestion
		case c, ok := <-provCh:
			if !ok {
				return
			}
```